Swiftz/Reader.swift

```diff
@@ -0,0 +1,174 @@
+//
+//  Reader.swift
+//  Swiftz
+//
+//  Created by Matthew Purland on 11/25/15.
+//  Copyright © 2015 TypeLift. All rights reserved.
+//
+
+/// A `Reader` monad with `R` for environment and `A` to represent the modified environment.
+public struct Reader<R, A> {
+    /// The function that modifies the environment
+    public let reader : R -> A
+    
+    init(_ reader : R -> A) {
+        self.reader = reader
+    }
+    
+    /// Transform the value returned by a reader
+    public func mapReader<B>(f : A -> B) -> Reader<R, B> {
+        return Reader<R, B>(f • runReader)
+    }
+    
+    /// Runs the reader and extracts the final value from it
+    public func runReader(environment : R) -> A {
+        return reader(environment)
+    }
+    
+    /// Executes a computation in a modified environment
+    public func local(f : R -> R) -> Reader<R, A> {
+        return Reader(reader • f)
+    }
+
+    /// Retrieves the monad environment
+    public func ask() -> R -> A {
+        return reader
+    }
+}
+
+/// Puts a value in a reader.
+public func returnReader<R, A>(a : A) -> Reader<R, A> {
+    return Reader<R, A>.pure(a)
+}
+
+/// Runs the reader and extracts the final value from it. This provides a global function for running a reader.
+public func runReader<R, A>(reader : R -> A) -> R -> A {
+    return runReader(Reader<R, A>(reader))
+}
+
+public func runReader<R, A>(reader : () -> R -> A) -> R -> A {
+    return runReader(reader())
+}
+
+public func runReader<R, A>(reader : Reader<R, A>) -> R -> A {
+    return reader.runReader
+}
+
+/// Retrieves the monad environment
+public func ask<A>() -> A -> A {
```

Comment:
It was intended to give ask functionality of the form A -> A to support runReader. 

```swift
let lengthResult3 = runReader(ask())(1234)
XCTAssert(lengthResult3 == 1234)
        
let lengthResult4 = runReader(ask)(1234)
XCTAssert(lengthResult4 == 1234)
```

I'm open to suggestions. How would you update the ```ask``` function to be a monadic action in the reader? So far I've run through a few different options, but each isn't working as I saw it should with runReader.

I've been working through some options, but none of these are right.

```swift
public static func ask<A>() -> (R -> A) -> Reader<R, A> {
    return { (f : R -> A) in
        return Reader<R, A>(f)
    }
}

public static func ask<A>() -> A -> Reader<R, A> {
    return { a in
        return Reader<R, A> { _ in a }
    }
}

public static func ask<A>(f : () -> A) -> Reader<A, A> {
    return Reader<A, A> { _ in f() }
}
```